File: mobile/ios/TerraTests/LiveTrackingTests.swift
import XCTest
@testable import Terra
final class LiveTrackingTests: XCTestCase {
    func testSpeedRejectsUnreliableSamplesAndExpires() {
        var speed = LiveSpeed(); speed.add(metersPerSecond: 2,accuracy: 0.3,time: 1000,maximum: 100)
        XCTAssertEqual(speed.value(now: 1000)!,7.2,accuracy: 0.01)
        speed.add(metersPerSecond: 50,accuracy: 10,time: 2000,maximum: 100)
        XCTAssertEqual(speed.value(now: 2000)!,7.2,accuracy: 0.01); XCTAssertNil(speed.value(now: 6000))
        speed.add(metersPerSecond: 0,accuracy: 0.5,time: 7000,maximum: 100); XCTAssertEqual(speed.value(now: 7000),0)
    }
    func testWrongModeStopsDiscoveryAndRecovers() {
        var guardrail = ModeSpeedGuard()
        for time in stride(from: 1000.0,through: 9000,by: 1000) { guardrail.update(kmh: 30,time: time,mode: .walk) }
        XCTAssertFalse(guardrail.allowsDiscovery); XCTAssertTrue(guardrail.warning)
        for time in stride(from: 10000.0,through: 13000,by: 1000) { guardrail.update(kmh: 5,time: time,mode: .walk) }
        XCTAssertTrue(guardrail.allowsDiscovery); XCTAssertFalse(guardrail.warning)
        guardrail.update(kmh: 30,time: 14000,mode: .bike); XCTAssertTrue(guardrail.allowsDiscovery)
    }
    func testExcludedPointsDoNotOpenMapAndSurviveSave() throws {
        let s = TrackSession(mode: .walk,startedAt: 0,endedAt: 10000,points: [TrackPoint(lat: 55,lng: 37,t: 1000,accuracy: 5,excludeDiscovery: true)],steps: 15)
        let restored = try JSONDecoder().decode(TrackSession.self,from: JSONEncoder().encode(s))
        XCTAssertTrue(Discovery.cells([restored]).isEmpty); XCTAssertEqual(restored.steps,15)
    }
    func testLegacyPhotoIsRetained() throws {
        let bytes = Data(#"{"id":"a","lat":55,"lng":37,"note":"Old place","photo":"old.jpg"}"#.utf8)
        let place = try JSONDecoder().decode(Place.self,from: bytes)
        XCTAssertEqual(place.attachments.first?.file,"old.jpg"); XCTAssertEqual(place.displayTitle,"Old place")
    }
}

File: mobile/android/app/src/test/java/app/terra/explore/LiveTrackingTests.kt
package app.terra.explore
import org.junit.Test
import org.junit.Assert.*
class LiveTrackingTests {
    @Test fun speedRejectsBadAccuracyAndExpires() {
        val speed=LiveSpeed(); speed.add(2.0,0.3,1000,100.0); assertEquals(7.2,speed.value(1000)!!,0.01)
        speed.add(50.0,10.0,2000,100.0); assertEquals(7.2,speed.value(2000)!!,0.01); assertNull(speed.value(6000))
        speed.add(0.0,0.5,7000,100.0); assertEquals(0.0,speed.value(7000)!!,0.0)
    }
    @Test fun wrongModeStopsDiscoveryAndRecovers() {
        val guard=ModeSpeedGuard(); for(t in 1000L..9000L step 1000) guard.update(30.0,t,Mode.walk)
        assertFalse(guard.allowsDiscovery); assertTrue(guard.warning)
        for(t in 10000L..13000L step 1000) guard.update(5.0,t,Mode.walk)
        assertTrue(guard.allowsDiscovery); assertFalse(guard.warning)
        guard.update(30.0,14000,Mode.bike); assertTrue(guard.allowsDiscovery)
    }
    @Test fun excludedPointsSurviveSave() {
        val session=Session(mode=Mode.walk,startedAt=0,endedAt=10000,points=mutableListOf(Point(55.0,37.0,1000,5.0,excludeDiscovery=true)),steps=15)
        val restored=Session.parse(session.json()); assertTrue(Discovery.cells(listOf(restored)).isEmpty()); assertEquals(15,restored.steps)
    }
}
